### PR TITLE
Replace Windows.Storage.Pickers with Common File Dialogs

### LINKED ISSUE #8957: Browse buttons crash when running as admin
The optional graphical settings editor worked properly when I ran Windows Terminal as a regular user. However, if I ran Windows Terminal as Administrator, the app would crash if I clicked the "Browse..." button for either "Command line" or "Starting directory".

_Originally posted by @huang63 in https://github.com/microsoft/terminal/issues/6800#issuecomment-769568274_

### PR DESCRIPTION
**Pull request by DHowett**
Using Pickers from an elevated application yields an
ERROR_ACCESS_DENIED. Of course it does: it was designed for the modern
app platform.

Using the common dialog infrastructure has some downsides¹, but it
doesn't crash and is just as flexible.

I've added some fun templated functions that help us with the
complexity.

Fixes #8957

¹You've got to use raw COM, and it runs in-proc instead of out-of-proc.

## Validation Steps Performed
I tested every picker.

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 1]
```diff
@@ -15,16 +15,71 @@ using namespace winrt::Windows::UI::Xaml::Data;
...
+
+    co_return winrt::hstring{ filePath.get() };
+}
+
+// Function Description:
+// - Helper that opens a file picker pre-seeded with image file types.
+static winrt::Windows::Foundation::IAsyncOperation<winrt::hstring> OpenImagePicker()
+{
+    static constexpr COMDLG_FILTERSPEC supportedImageFileTypes[] = {
+        { L"All Supported Bitmap Types (*.jpg, *.jpeg, *.png, *.bmp, *.gif, *.tiff, *.ico)", L"*.jpg;*.jpeg;*.png;*.bmp;*.gif;*.tiff;*.ico" },
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:64]: The previous dialog didn't even have this bit of text, so I'm going to elave this unlocalized for now.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 2]
```diff
@@ -15,16 +15,71 @@ using namespace winrt::Windows::UI::Xaml::Data;
...
+        { L"All Files (*.*)", L"*.*" }
+    };
+
+    static constexpr winrt::guid clientGuidImagePicker{ 0x55675F54, 0x74A1, 0x4552, { 0xA3, 0x9D, 0x94, 0xAE, 0x85, 0xD8, 0xF2, 0x7A } };
+    return OpenFilePicker([](auto&& dialog) {
+        THROW_IF_FAILED(dialog->SetClientGuid(clientGuidImagePicker));
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:70]: the client GUID is how it remembers the location, so we can have it remember different locations for different types of box!

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 3]
```diff
@@ -15,16 +15,71 @@ using namespace winrt::Windows::UI::Xaml::Data;
...
 };
 
+// Function Description:
+// - This function presents a File Open "common dialog" and returns its selected file asynchronously.
+// Parameters:
+// - customize: A lambda that receives an IFileDialog* to customize.
+// Return value:
+// (async) path to the selected item.
+template<typename TLambda>
+static winrt::Windows::Foundation::IAsyncOperation<winrt::hstring> OpenFilePicker(TLambda&& customize)
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:32]: Should we put a comment here linking 8957 that says why we shouldn't use `Windows::Storage::Pickers::FileOpenPicker`?

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:64]: follow-up issue #?

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:64]: (same with "all files" below)

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 4]
```diff
@@ -582,74 +637,74 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
         }
     }
 
     fire_and_forget Profiles::Commandline_Click(IInspectable const&, RoutedEventArgs const&)
     {
         auto lifetime = get_strong();
 
-        FileOpenPicker picker;
+        static constexpr COMDLG_FILTERSPEC supportedFileTypes[] = {
+            { L"Executable FIles (*.exe, *.cmd, *.bat)", L"*.exe;*.cmd;*.bat" },
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:663]:
```suggestion
            { L"Executable Files (*.exe, *.cmd, *.bat)", L"*.exe;*.cmd;*.bat" },
```

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 5]
```diff
@@ -582,74 +637,74 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
-            _State.Profile().Commandline(file.Path());
+            _State.Profile().Commandline(path);
         }
     }
 
     fire_and_forget Profiles::StartingDirectory_Click(IInspectable const&, RoutedEventArgs const&)
     {
         auto lifetime = get_strong();
-        FolderPicker picker;
-        _State.WindowRoot().TryPropagateHostingWindow(picker); // if we don't do this, there's no HWND for it to attach to
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:645]: Can we revert `TryPropagateHostingWindow`? PR #8391

**Review by carlos-zamora — APPROVED** [review]
Small things. I don't want to hold up this PR, so I'm ok with follow-up work items on this.

**Comment by carlos-zamora** [comment]
Also, you tested this in as Admin, right?

**Comment by DHowett** [comment]
> Also, you tested this in as Admin, right?

It would not have been a worthwhile change had I not tested it 😉

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 5 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:645]: Happy to revert it for 1.9, but not for 1.8 or 1.7.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 3 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:32]: I guess. We can also simply remain vigilant on PRs that attempt to add it back :p

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 1 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:64]: It is my suspicion that this text does not _need_ to be localized, as it is a file picker that filters file types and the `*.bmp, *.jpg` is directly understandable by anyone who might try to read the string.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 6]
```diff
@@ -582,74 +637,74 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
-        picker.ViewMode(PickerViewMode::Thumbnail);
-        picker.SuggestedStartLocation(PickerLocationId::PicturesLibrary);
-
-        // Converted into a BitmapImage. This list of supported image file formats is from BitmapImage documentation
-        // https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Media.Imaging.BitmapImage?view=winrt-19041#remarks
-        picker.FileTypeFilter().ReplaceAll({ L".jpg", L".jpeg", L".png", L".bmp", L".gif", L".tiff", L".ico" });
-
-        StorageFile file = co_await picker.PickSingleFileAsync();
-        if (file != nullptr)
+        auto file = co_await OpenImagePicker();
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:640]: Do we need to hop to the background thread before or during this `co_await`? Or does the IFileDialog do that automagically?

**Review by zadjii-msft — COMMENTED** [review]
I'm only holding the approval to make sure we don't end up blocking the UI thread entirely (or if we do, then it's intentional)

**Comment by zadjii-msft** [comment]
Morbid curiosity: will this end up changing the CWD of `wt.exe`, such that commandlines with `"startingDirectory": "."` set would behave differently after using this dialog?

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 6 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:640]: It's actually the co_await that splits off to the right thread! However: we actually do want to block the UI thread- the dialog is naturally a UI thing, and we don't want the user navigating to another profile page while the dialog is up. Yes, that sucks, but I don't believe this is a difference from Pickers.

**Comment by DHowett** [comment]
CWD: we actually have disabled the current directory change feature that's built into the dialogs using FOS_NOCHANGEsomething

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:640]: Oh, wait really? I always assumed that until you `resume_*` then you're still on the main thread.

I guess blocking the UI thread is in fact the right behavior, I guess I just didn't want the window to do the "Window unresponsive" thing, but I guess Windows knows how to deal with that already.

**Review by zadjii-msft — APPROVED** [review]

**Pull request merged**